Snapshot APIs, Create Models, PUT Response to 201 added

PreetPRG opened the pull request:
## ARM API Information (Control Plane)

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow.
</i>
Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [X] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  April-2023
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  March-2023
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [X] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [X] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [X] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [X] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[X] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [X] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

PreetPRG committed:
Snapshot APIs, Create Models, PUT Response to 201 added

mentat9 commented on specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:
```diff
@@ -1008,6 +1012,297 @@
+        "operationId": "Snapshots_ListByVolumeGroup",
+        "parameters": [
+          {
+            "$ref": "../../../../../common-types/resource-management/v2/types.json#/parameters/SubscriptionIdParameter"
```
>v2 [](http://example.com/codeflow?start=69&length=2)

Recommend latest common types version (v5). Comment applies throughout this PR.

mentat9 commented on specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:
```diff
@@ -1008,6 +1012,297 @@
+            "$ref": "../../../../../common-types/resource-management/v2/types.json#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "$ref": "#/parameters/ResourceGroupNameParameter"
```
>#/parameters/ResourceGroupNameParameter [](http://example.com/codeflow?start=21&length=39)

Recommend getting this from common types. Comment applies throughout this PR.

mentat9 commented on specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:
```diff
@@ -1008,6 +1012,297 @@
+        ],
+        "responses": {
+          "200": {
+            "description": "OK -- Returned when the create operation is successfully completed.",
```
>OK -- Returned when the create operation is successfully completed [](http://example.com/codeflow?start=28&length=66)

200 is the response in the update case

mentat9 commented on specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:
```diff
@@ -1008,6 +1012,297 @@
+            }
+          },
+          "201": {
+            "description": "Created -- Resource Created; operation will complete asynchronously.",
```
>Created -- Resource Created; operation will complete asynchronously. [](http://example.com/codeflow?start=28&length=68)

201 is the response in the create case

PreetPRG replied:
I didn't get you, we are responding with 201 for PUT operations, any other change?

mentat9 commented on specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:
```diff
@@ -1960,6 +2312,119 @@
+      },
+      "allOf": [
+        {
+          "$ref": "#/definitions/Resource"
```
>"#/definitions/Resource" [](http://example.com/codeflow?start=18&length=24)

Get this from v5 common types. Same comment for SnapshotCreateParameter.

TimLovellSmith replied:
And usually its preferable to inherit specifically ProxyResource or TrackedResource than Resource.

mentat9 commented on specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:
```diff
@@ -1960,6 +2312,119 @@
+          "readOnly": true,
+          "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData"
+        }
+      },
```
Use the Resource definition from common types and remove this. Same comment for SnapshotCreateParameter.

PreetPRG replied:
The resource definition differ from definition mentioned in our swagger. The tags field is mentioned under tracked resource in common types while we have it in Resource model.

TimLovellSmith replied:
I don't think that is a very significant difference. And declaring Tags in 'Resource' is technically not really correct because only tracked resources always support Tags.

TimLovellSmith replied:
@tadelesh From an sdk point of view, what do you think they should do?

TimLovellSmith replied:
So @mentat9 is mentioning that Snapshot never should have had 'Tags' anyway, if it provides that functionality, it should be called something else. Has it already shipped with them though? Since it used to inherit them through the (incorrect) Resource definition?

TimLovellSmith replied:
@rkmanda @mentat9 Any further thoughts on whether the 'no tags on proxy resources' rule merits any breaking changes?

mentat9 commented on specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:
```diff
@@ -1960,6 +2312,119 @@
+        }
+      ]
+    },
+    "SnapshotCreateParameter": {
```
>SnapshotCreateParameter [](http://example.com/codeflow?start=5&length=23)

How is this different from definitions/Snapshot? Recommend unifying on Snapshot: they should be the same anyway.

PreetPRG replied:
They both are different, SnapshotCreateParameter Model has required = CreationData while Snapshot Model has nothing as required.
The idea to have a separate model for Create operation was to take care of required properties.

TimLovellSmith replied:
Would it be incrementally better to inherit allOf Snapshot, but just add some requiredness onto that?

mentat9 commented on specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:
```diff
@@ -1008,6 +1012,297 @@
+          "default": {
+            "description": "Error from the ElasticSan Resource Provider",
+            "schema": {
+              "$ref": "#/definitions/Error"
```
>#/definitions/Error [](http://example.com/codeflow?start=23&length=19)

Recommend using ErrorResponse from common types. Comment applies throughout this PR.

mentat9 commented on specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:
```diff
@@ -1960,6 +2312,119 @@
+      "type": "object",
+      "description": "Snapshot request.",
+      "properties": {
+        "tags": {
```
>tags [](http://example.com/codeflow?start=9&length=4)

Snapshot seems to be a proxy type. Don't put tags on a proxy type: this is an ARM anti-pattern.

PPrahalad replied:
Thanks @mentat9 .

Quick follow-up : how do we provide "tag-like" feature for proxy resources. In our case, we are expecting to have a large number of snapshots and hence we want to provide the user a way of classifying snapshots with appropriate tags.

PreetPRG replied:
@mentat9 / @tadelesh can you please give us clarity on this too? Thanks

mentat9 replied:
@PreetPRG - It's OK to provide that kind of feature, but definitely don't call it "tags". Make sure it's obvious to clients that it's not the same as tags provided by ARM for tracked resources. It's common for APIs to provide a "metadata" or "category" property that clients can populate (or partially populate). You can find examples in other swaggers.

PreetPRG commented:
Hi @tadelesh, ElasticSan is planning to go for GA soon and for the same we are having a new API version.
Things added in new API version:
1) Added new resource i.e. volumeSnapshot, which can have source as volume.
Changed:
i first submitted a PR and got couple of LintErrors and model validation errors, so i have resolved most of them in this PR. Changes:
2) PUT response from 202 to 201.
3) Request model body during create call, made a separate model for create and response.

In Doubt:
4) There are some LintErrors still in this PR, 2 errors are common cross the spec:
      a)A 202 response should include an Location response header.
      b) An async DELETE operation is tracked via Azure-AsyncOperation header. Set 'final-state-via' property to 'location' on 'x-ms-long-running-operation-options'

a) From Service side we are including location response header while responding with 202. Should i add it in swagger spec too?
b) why we need to change 'x-ms-long-running-operation-options' ->  'final-state-via' property to 'location', in preview swagger we had 'final-state-via' property to 'azure-sync-operation' ?

Thanks

PreetPRG committed:
updated v2 to v5, updated ResourceGroupNameParameter

PreetPRG committed:
Revert " updated v2 to v5, updated ResourceGroupNameParameter"

This reverts commit 281e9ba747cd45cb291ac97d81d8b3f192ec1e5a.

PreetPRG committed:
Updated v2 to v5

PreetPRG committed:
added with prettier check

PreetPRG commented:
Hi @mentat9, somehow updating common type's version from v2 to v5 is breaking the pipeline and PR is getting labelled with CI-FixRequiredOnFailure. Any reason for this? I am just updating the version here.

mentat9 commented:
> Hi @mentat9, somehow updating common type's version from v2 to v5 is breaking the pipeline and PR is getting labelled with CI-FixRequiredOnFailure. Any reason for this? I am just updating the version here.

I don't know what that label means. Note that PR validation/approval is owned by the Azure SDK team. @tadelesh, any idea?

PreetPRG committed:
updated v2 to v5 after draft check

PreetPRG commented:
> > Hi @mentat9, somehow updating common type's version from v2 to v5 is breaking the pipeline and PR is getting labelled with CI-FixRequiredOnFailure. Any reason for this? I am just updating the version here.
> 
> I don't know what that label means. Note that PR validation/approval is owned by the Azure SDK team. @tadelesh, any idea?

@mentat9/ @tadelesh  I created a draft PR with the same content, strangely that does not get tagged with CIFixRequiredOnFailure. Why such? Is this a Github bug? What should we do to get unblocked from this?
Link to draft PR: https://github.com/Azure/azure-rest-api-specs/pull/23073

tadelesh commented:
> > > Hi @mentat9, somehow updating common type's version from v2 to v5 is breaking the pipeline and PR is getting labelled with CI-FixRequiredOnFailure. Any reason for this? I am just updating the version here.
> > 
> > 
> > I don't know what that label means. Note that PR validation/approval is owned by the Azure SDK team. @tadelesh, any idea?
> 
> @mentat9/ @tadelesh I created a draft PR with the same content, strangely that does not get tagged with CIFixRequiredOnFailure. Why such? Is this a Github bug? What should we do to get unblocked from this? Link to draft PR: #23073

@PreetPRG I don't familiar with the tooling logic and I'm asking tooling team to help to take a look. The `x-ms-long-running-operation-options` looks reasonable to me.

PreetPRG commented:
> > > > Hi @mentat9, somehow updating common type's version from v2 to v5 is breaking the pipeline and PR is getting labelled with CI-FixRequiredOnFailure. Any reason for this? I am just updating the version here.
> > > 
> > > 
> > > I don't know what that label means. Note that PR validation/approval is owned by the Azure SDK team. @tadelesh, any idea?
> > 
> > 
> > @mentat9/ @tadelesh I created a draft PR with the same content, strangely that does not get tagged with CIFixRequiredOnFailure. Why such? Is this a Github bug? What should we do to get unblocked from this? Link to draft PR: #23073
> 
> @PreetPRG I don't familiar with the tooling logic and I'm asking tooling team to help to take a look. The `x-ms-long-running-operation-options` looks reasonable to me.

Thanks @tadelesh , should I go ahead and make more changes as reviewed by @mentat9 ? If possible can you also add me to the thread with tooling team?

tadelesh commented:
@PreetPRG I was informed that this is a new ARM rule that asynchronized DELETE response headers MUST include a Location header that points to a URL where the ongoing operation can be monitored, you could refer [this](https://github.com/Azure/azure-openapi-validator/blob/main/docs/delete-operation-async-response-validation.md). I wonder if @mentat9 knows more.

PreetPRG commented:
/azp run

PreetPRG committed:
resolved Lint err, added volumeName, updated RGName, Error to point commonTypes

PreetPRG commented:
/azp run

PreetPRG commented:
Hi @mentat9 and @tadelesh, the CI-FixedRequiredOnFailure tag issue is resolved. Also, i have made changes as suggested by @mentat9. Can you Please review it? Thanks

tadelesh commented:
1. Since you have change to use the common type of `ErrorResponse`, you need to remove the previous self-defined one in your swagger to prevent duplicate schema issue of SDK generation.
2. You need to follow the guide to go through breaking change review.

PreetPRG committed:
Removed Error Model

PreetPRG commented:
Thanks @tadelesh, i have removed the ErrorResponse-related models from swagger.

Follow up questions:
1)  How do we provide "tag-like" feature for proxy resources. In our case, we are expecting to have a large number of snapshots and hence we want to provide the user a way of classifying snapshots with appropriate tags.
2) What do you mean by guideline to follow for breaking change review?

TimLovellSmith commented on specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:
```diff
@@ -1960,6 +2281,124 @@
+          "description": "Properties of Volume Snapshot.",
+          "$ref": "#/definitions/SnapshotProperties"
+        },
+        "systemData": {
```
>systemData [](http://example.com/codeflow?start=9&length=10)

systemData would be included automatically if you inherit ProxyResource from common types

TimLovellSmith commented:
>     "SubscriptionIdParameter": {

These are in common-types so you could reference them instead

[https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v5/types.json#L630](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v5/types.json#L630)

---
Refers to: specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:1344 in cf1ed6b. [](commit_id = cf1ed6b108f577980fd184794c5368f239a0adc6, deletion_comment = False)

TimLovellSmith commented on specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:
```diff
@@ -1017,15 +1347,17 @@
+      "minLength": 1,
+      "x-ms-parameter-location": "client"
     },
     "ApiVersionParameter": {
```
>ApiVersionParameter [](http://example.com/codeflow?start=5&length=19)

[https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v5/types.json#L630](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v5/types.json#L630)

TimLovellSmith commented on specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:
```diff
@@ -1017,15 +1347,17 @@
+      "minLength": 1,
+      "x-ms-parameter-location": "client"
     },
     "ResourceGroupNameParameter": {
```
>ResourceGroupNameParameter [](http://example.com/codeflow?start=5&length=26)

[https://github.com/Azure/azure-rest-api-specs/blob/a5c02f03ec242a2acef309bbdfd4734c100f1839/specification/common-types/resource-management/v5/types.json#L646](https://github.com/Azure/azure-rest-api-specs/blob/a5c02f03ec242a2acef309bbdfd4734c100f1839/specification/common-types/resource-management/v5/types.json#L646)

PreetPRG committed:
updated createModels to inherit actual model

PreetPRG committed:
Removed SubscriptionID, APIVersion, RGName parameter from swagger

PreetPRG commented:
Thanks @TimLovellSmith  , made changes as per your comments.

TimLovellSmith commented on specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:
```diff
@@ -1150,9 +1464,6 @@
     },
```
Do you need to customize these? I guess its probably too late now from point of view of breaking changes. There are standard typedefs for these in common-types though. Maybe you could inherit them or something. Not really important, consdier doing it later?

TimLovellSmith commented on specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:
```diff
@@ -1850,62 +2174,6 @@
       "description": "The resource model definition.",
       "x-ms-azure-resource": true
     },
```
You really shouldn't be defining TrackedResource yourself, you should just inherit the one from common-types instead.

TimLovellSmith commented on specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:
```diff
@@ -1185,7 +1505,7 @@
         "systemData": {
           "description": "Resource metadata required by ARM RPC",
```
You won't need to declare this if you inherit TrackedResource from common-types/v3 or newer

TimLovellSmith commented on specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:
```diff
@@ -1512,7 +1832,7 @@
         "systemData": {
```
>systemData [](http://example.com/codeflow?start=9&length=10)

You won't need to declare this systemData, if you inherit TrackedResource from common-types/v3 or newer, which includes it

PreetPRG committed:
Removed Import Enum

TimLovellSmith commented:
>           ],

You turned it into an enum with a single value, which is a little unusual. Do you still need this property?
What does createSource "None" mean, anyway?

---
Refers to: specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:2045 in 7ecac55. [](commit_id = 7ecac55d2038b1237ea92145935e6539572ae783, deletion_comment = False)

tadelesh commented on specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json:
```diff
@@ -746,12 +770,12 @@
             "required": true,
             "schema": {
-              "$ref": "#/definitions/Volume"
+              "$ref": "#/definitions/VolumeCreateParameter"
```
All such kinds of changing parameter ref will be breaking changes for SDKS. @mikekistler , I wonder whether they are reviewed by board?

tadelesh replied:
> Had this been the only change, I don't think it would be flagged for breaking change review because "ReferenceRedirection" is only reported as a warning and not an error. But the pipeline should detect breaking changes in the SDKs and flag those for review.

Ray has helped to create a request to add it to breaking change tool. Could you help to re-review and approve such breaking changes in this PR?

mikekistler replied:
Please send email to @azbreakchangereview with a link to the ADO item.

tadelesh replied:
@PreetPRG could you do the breaking change review again?
